### PR TITLE
CUMULUS-2462: Implement Reingest and Bulk Reingest to select one workflow from multiple executions

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-2462: Dashboard: Implement Reingest and Bulk Reingest to select one workflow from multiple executions](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2462)

## Changes

* Add actions and reducers for get/clear/error granules workflows.
* Reingest a granule from granule details page is updated to select a workflow execution arn for reingest.
* Batch Reingest and Bulk Reingest Actions are updated to select the workflowName for reingest.
* list of selected granule object {granuleId, collectionId} is passed to `reingest BatchAsyncCommands`, and `BulkGranuleModal`, so they can be used to get list of workflow names for the selected granules.   The actions (reingest, execute, bulkGranule etc.) still take list of granuleIds for payload
* For batch and bulk reingest, if `workflowName` is selected, it is added to reingest payload.
* For reingest an individual granule, if executionArn of the selected workflow is passed to reingest payload.
* ErrorReport is displayed in reingest modals  if there is error when getting granule workflow/execution list.
* Added/updated cypress tests for single, batch, bulk granule reingest.
* Add `seedGranulesExecutions` method to add executions to granule
* Updated executions_spec to use local api for getting granule executions list, removed executions-list.json 

### Test agains a Local API
* Use granules `MOD09GQ.A9344328.K9yI3O.006.4625818663028` and `MOD09GQ.A2417309.YZ9tCV.006.4640974889044`, and these two granules have multiple executions

### Test against a cumulus API
* Ingest 2-3 granules into your cumulus stack, and do NOT cleanup the collections, providers and test data after ingest
* Add a workflow execution to these granules by applying HelloWorldWorkflow to these granules. Select these 2 granules, `Granule Actions-> Execute ->select HelloWorldWorkflow`, and verify the workflow executions are Completed from Executions tab.  Note the granules stay in 'Running' because the HelloWorldWorkflow's output payload doesn't have granule.

* Reingest a single granule with selected workflow.  Go to an individual granule page, select `Reingest` from the `Options` dropdown, and select a workflow from the list, and `Confirm`.  Verify that the selected workflow `executionArn` is passed to reingest endpoint.
* Reingest a single granule w/o select a workflow.  Go to an individual granule page, select `Reingest` from the `Options` dropdown, and `Confirm`. Verify the default/latest workflow is executed.

* Batch Reingest Granules with selected workflow.  Select these granules, `Granule Actions -> Reingest`, verify that both `HelloWorldWorkflow` and `IngestAndPublishWorkflow` are in the dropdown list.  Select `IngestAndPublishWorkflow` which is not the latest execution, and `Confirm`.  Verify that the selected workflow `workflowName` is passed to reingest endpoint.
* Batch Reingest Granules w/o select a workflow.  Select one granule, `Granule Actions -> Reingest`, verify that all workflows for this granule are listed in the dropdown, and do not select a workflow, the lasted workflow will be executed.

* Bulk Reingest Granules with selected workflow.  Select these granules, `Granule Actions -> Run Bulk Granules -> Run Bulk Reingest`, verify that both `HelloWorldWorkflow` and `IngestAndPublishWorkflow` are in the dropdown list.  Select `HelloWorldWorkflow` which is not the latest execution, and `Confirm`.  Verify that the selected workflow `workflowName` is passed to reingest endpoint.
* Bulk Reingest Granules w/o select a workflow.  Select one granule, `Granule Actions -> Run Bulk Granules -> Run Bulk Reingest`, verify that all workflows for this granule are listed in the dropdown, and do not select a workflow, the lasted workflow will be executed.

* If there is issue getting workflows for granules, the error should be displayed above the `Select Workflow` dropdown.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [x] Integration tests

### CODE REVIEW

File: app/src/js/components/Section/section.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ export const section = ({
 }) => {
   const borderMod = border ? '--border' : '';
   const sizeMod = `--${size}`;
-  const id = sectionId || false;
+  const id = sectionId || 'false';
```
Comment (jennyhliu) [inline app/src/js/components/Section/section.js:18]: Fix error in the JavaScript Console:
react_devtools_backend.js:2574 Warning: Received `false` for a non-boolean attribute `id`.

**Comment by jjmccoy** [comment]
UI/UX: I will be reviewing for interface styling and interaction. I suggest another reviewer for code, different APIs, and testing against a large dataset.

File: app/src/js/components/AsyncCommands/AsyncCommands.js [hunk 2]
```diff
@@ -138,7 +138,7 @@ AsyncCommand.propTypes = {
   disabled: PropTypes.bool,
   element: PropTypes.string,
   confirmAction: PropTypes.bool,
-  confirmText: PropTypes.string,
+  confirmText: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
```
Comment (jennyhliu) [inline app/src/js/components/AsyncCommands/AsyncCommands.js:141]: This it to fix the warning message in JavaScript console:
react_devtools_backend.js:4049 Warning: Failed prop type: Invalid prop `confirmText` of type `array` supplied to `AsyncCommand`, expected `string`.
    in AsyncCommand (created by DropdownAsync)
    in DropdownAsync (created by Granule)
    in section (created by Granule)
    in div (created by Granule)
    in Granule (created by Connect(Granule))
    in Connect(Granule) (created by Context.Consumer)
    in withRouter(Connect(Granule)) (created by Context.Consumer)

**Comment by jjmccoy** [comment]
@jennyhliu Things look fine on the UI/UX so far. I'm running into an issue (same thing I had with Anthony & Danielle's tickets with the array .flat function error -- I'm seeing in Chrome & Firefox) that I'm troubleshooting in order to see the associated workorders in the "Run Bulk Granules" workflow. I put in a ticket ([CUMULUS-2699](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2699)) for a UI change from the modal to a page and that should help with styling and interaction concerns.

**Comment by jennyhliu** [comment]
@jjmccoy Do you use node v12.18.0 for  localAPI? The `flat` function isn't supported in older version of node [link](https://node.green/#ES2019-features-Array-prototype--flat--flatMap-)

**Comment by jjmccoy** [comment]
> @jjmccoy Do you use node v12.18.0 for localAPI? The `flat` function isn't supported in older version of node [link](https://node.green/#ES2019-features-Array-prototype--flat--flatMap-)

Yes. When I use nvm I get this:
```
nvmrc' with version <12.18.0>
Now using node v12.18.0 (npm v6.14.4)
```
And then I `npm install`. I did the `npm ci` instructions in your README update as well.

File: cypress/integration/granules_spec.js [hunk 3]
```diff
@@ -376,6 +376,27 @@ describe('Dashboard Granules Page', () => {
...
+      );
+      cy.intercept('GET', `/granules/${granuleId}*`).as('getGranule');
+
+      cy.visit(`/granules/granule/${granuleId}`);
+      cy.wait('@getGranule');
+      cy.get('.heading--large').should('have.text', `Granule: ${granuleId}`);
+      cy.contains('button', 'Options').click();
+      cy.get('.dropdown__menu').contains('Reingest').click();
+      cy.get('.modal-body .form__dropdown .dropdown__element input').as('workflow-input');
+      cy.get('@workflow-input').click({ force: true }).type('IngestAndPublish').type('{enter}');
```
Comment (flamingbear) [inline cypress/integration/granules_spec.js:393]: Is this a good test?  I can type "anything" in the box and then hit confirm/select and get a success.  I'm 99% sure it's just not selecting anything when I do that.

Comment (flamingbear) [reply cypress/integration/granules_spec.js:393]: I don't know if that's a bug or feature. 🤷

**Review by flamingbear — CHANGES_REQUESTED** [review]
Wow, this is a lot of great work.  I only had a couple of comments, and a question that I'd like you to walk me through. 

I did run though all of the tests on the PR and verified that everything worked as expected, so I don't have any functional problems with the code.

File: app/src/js/components/ListActions/ListActions.js [hunk 4]
```diff
@@ -37,7 +37,8 @@ const ListActions = ({
       const { Component, text } = item;
       return (
         <React.Fragment key={text || index}>
-          {Component && React.cloneElement(Component, { selected })}
+          {/* 'selected' prop of BulkGranuleModal is array of objects, selected here is array of string */}
+          {Component && React.cloneElement(Component, Component.props.selected === undefined && { selected })}
```
[the remark this answers is not in the document]
Comment (jennyhliu) [reply app/src/js/components/ListActions/ListActions.js:43]: Updated comments.  Basically, for BulkGranule actions, we passed in selected which is array of {collectionId,granuleId}, and other actions may not have 'selected' property.
[ea523d8](https://github.com/nasa/cumulus-dashboard/pull/986/commits/ea523d875c6ca7cebc3f7f5119e440c5e953042e)

File: cypress/integration/granules_spec.js (hunk 3 above)
Comment (jennyhliu) [reply cypress/integration/granules_spec.js:393]: Updated tests to verify that a correct value is selected from dropdown.
[d816123](https://github.com/nasa/cumulus-dashboard/pull/986/commits/d816123c40054dc268dce769b0cc0bbffcdbfdf7)

**Review by flamingbear — APPROVED** [review]

**Comment by flamingbear** [comment]
@jennyhliu sorry I got distracted. I've approved your changes for the technical working side of things.

**Comment by jjmccoy** [comment]
UI/UX review: Thanks @jennyhliu for that quick styling change. LGTG!

**Pull request merged**